### ISSUE TITLE
Difference between convolution in speechbrain and pyroomacoustics

### ISSUE THREAD

**Issue by coreeey** [LCAV/pyroomacoustics#389]
Is there a difference between adding reverberation using Speechbrain and using pyroomacoustics? Can the operation of adding reverberation in Speechbrain be used as a substitute?
`def convolve1d(
    waveform,
    kernel,
    padding=0,
    pad_type="constant",
    stride=1,
    groups=1,
    use_fft=False,
    rotation_index=0,
):
    """Use torch.nn.functional to perform 1d padding and conv.

    Arguments
    ---------
    waveform : tensor
        The tensor to perform operations on.
    kernel : tensor
        The filter to apply during convolution.
    padding : int or tuple
        The padding (pad_left, pad_right) to apply.
        If an integer is passed instead, this is passed
        to the conv1d function and pad_type is ignored.
    pad_type : str
        The type of padding to use. Passed directly to
        `torch.nn.functional.pad`, see PyTorch documentation
        for available options.
    stride : int
        The number of units to move each time convolution is applied.
        Passed to conv1d. Has no effect if `use_fft` is True.
    groups : int
        This option is passed to `conv1d` to split the input into groups for
        convolution. Input channels should be divisible by the number of groups.
    use_fft : bool
        When `use_fft` is passed `True`, then compute the convolution in the
        spectral domain using complex multiply. This is more efficient on CPU
        when the size of the kernel is large (e.g. reverberation). WARNING:
        Without padding, circular convolution occurs. This makes little
        difference in the case of reverberation, but may make more difference
        with different kernels.
    rotation_index : int
        This option only applies if `use_fft` is true. If so, the kernel is
        rolled by this amount before convolution to shift the output location.

    Returns
    -------
    The convolved waveform.

    Example
    -------
    >>> from speechbrain.dataio.dataio import read_audio
    >>> signal = read_audio('tests/samples/single-mic/example1.wav')
    >>> signal = signal.unsqueeze(0).unsqueeze(2)
    >>> kernel = torch.rand(1, 10, 1)
    >>> signal = convolve1d(signal, kernel, padding=(9, 0))
    """
    if len(waveform.shape) != 3:
        raise ValueError("Convolve1D expects a 3-dimensional tensor")

    # Move time dimension last, which pad and fft and conv expect.
    waveform = waveform.transpose(2, 1)
    kernel = kernel.transpose(2, 1)

    # Padding can be a tuple (left_pad, right_pad) or an int
    if isinstance(padding, tuple):
        waveform = torch.nn.functional.pad(
            input=waveform, pad=padding, mode=pad_type
        )

    # This approach uses FFT, which is more efficient if the kernel is large
    if use_fft:

        # Pad kernel to same length as signal, ensuring correct alignment
        zero_length = waveform.size(-1) - kernel.size(-1)

        # Handle case where signal is shorter
        if zero_length < 0:
            kernel = kernel[..., :zero_length]
            zero_length = 0

        # Perform rotation to ensure alignment
        zeros = torch.zeros(
            kernel.size(0), kernel.size(1), zero_length, device=kernel.device
        )
        after_index = kernel[..., rotation_index:]
        before_index = kernel[..., :rotation_index]
        kernel = torch.cat((after_index, zeros, before_index), dim=-1)

        # Multiply in frequency domain to convolve in time domain
        if version.parse(torch.__version__) > version.parse("1.6.0"):
            import torch.fft as fft

            result = fft.rfft(waveform) * fft.rfft(kernel)
            convolved = fft.irfft(result, n=waveform.size(-1))
        else:
            f_signal = torch.rfft(waveform, 1)
            f_kernel = torch.rfft(kernel, 1)
            sig_real, sig_imag = f_signal.unbind(-1)
            ker_real, ker_imag = f_kernel.unbind(-1)
            f_result = torch.stack(
                [
                    sig_real * ker_real - sig_imag * ker_imag,
                    sig_real * ker_imag + sig_imag * ker_real,
                ],
                dim=-1,
            )
            convolved = torch.irfft(
                f_result, 1, signal_sizes=[waveform.size(-1)]
            )

    # Use the implementation given by torch, which should be efficient on GPU
    else:
        convolved = torch.nn.functional.conv1d(
            input=waveform,
            weight=kernel,
            stride=stride,
            groups=groups,
            padding=padding if not isinstance(padding, tuple) else 0,
        )

    # Return time dimension to the second dimension.
    return convolved.transpose(2, 1)

def reverberate(waveforms, rir_waveform, rescale_amp="avg"):
    """
    General function to contaminate a given signal with reverberation given a
    Room Impulse Response (RIR).
    It performs convolution between RIR and signal, but without changing
    the original amplitude of the signal.

    Arguments
    ---------
    waveforms : tensor
        The waveforms to normalize.
        Shape should be `[batch, time]` or `[batch, time, channels]`.
    rir_waveform : tensor
        RIR tensor, shape should be [time, channels].
    rescale_amp : str
        Whether reverberated signal is rescaled (None) and with respect either
        to original signal "peak" amplitude or "avg" average amplitude.
        Choose between [None, "avg", "peak"].

    Returns
    -------
    waveforms: tensor
        Reverberated signal.
    """
    orig_shape = waveforms.shape

    if len(waveforms.shape) > 3 or len(rir_waveform.shape) > 3:
        raise NotImplementedError

    # if inputs are mono tensors we reshape to 1, samples
    if len(waveforms.shape) == 1:
        waveforms = waveforms.unsqueeze(0).unsqueeze(-1)
    elif len(waveforms.shape) == 2:
        waveforms = waveforms.unsqueeze(-1)

    if len(rir_waveform.shape) == 1:  # convolve1d expects a 3d tensor !
        rir_waveform = rir_waveform.unsqueeze(0).unsqueeze(-1)
    elif len(rir_waveform.shape) == 2:
        rir_waveform = rir_waveform.unsqueeze(-1)

    # Compute the average amplitude of the clean
    orig_amplitude = compute_amplitude(
        waveforms, waveforms.size(1), rescale_amp
    )

    # Compute index of the direct signal, so we can preserve alignment
    value_max, direct_index = rir_waveform.abs().max(axis=1, keepdim=True)

    # Making sure the max is always positive (if not, flip)
    # mask = torch.logical_and(rir_waveform == value_max,  rir_waveform < 0)
    # rir_waveform[mask] = -rir_waveform[mask]

    # Use FFT to compute convolution, because of long reverberation filter
    waveforms = convolve1d(
        waveform=waveforms,
        kernel=rir_waveform,
        use_fft=True,
        rotation_index=direct_index,
    )

    # Rescale to the peak amplitude of the clean waveform
    waveforms = rescale(
        waveforms, waveforms.size(1), orig_amplitude, rescale_amp
    )

    if len(orig_shape) == 1:
        waveforms = waveforms.squeeze(0).squeeze(-1)
    if len(orig_shape) == 2:
        waveforms = waveforms.squeeze(-1)

    return waveforms`
    
    `def convolve(sources, rir, snr=None, fs=16000, sigma2_awgn=None):
    from scipy.signal import fftconvolve

    max_len_rir = len(rir)
    max_sig_len = len(sources)
    L = int(max_len_rir) + int(max_sig_len) - 1
    if L % 2 == 1:
        L += 1

    # the array that will receive all the signals
    premix_signals = np.zeros((1, 1, L))

    # compute the signal at every microphone in the array

    sig = sources
    d = int(np.floor(0 * fs))
    h = rir
    premix_signals[0, 0, d: d + len(sig) + len(h) - 1] += fftconvolve(
        h, sig
    )

    if snr is not None:
        # Normalize all signals so that
        denom = np.std(premix_signals[:, 0, :], axis=1)
        premix_signals /= denom[:, None, None]
        signals = np.sum(premix_signals, axis=0)

        # Compute the variance of the microphone noise
        sigma2_awgn = 10 ** (-snr / 10) * 1

    else:
        signals = np.sum(premix_signals, axis=0)

    # add white gaussian noise if necessary
    if sigma2_awgn is not None:
        signals += np.random.normal(0.0, np.sqrt(sigma2_awgn), signals.shape)

    # record the signals in the microphones
    return signals`

**Comment by coreeey** [comment]
here is the link of speechbrain
https://github.com/speechbrain/speechbrain/blob/b18646d80c7d5081a99a32f2a3e44cebe2a8249f/speechbrain/processing/signal_processing.py#L196

**Comment by fakufaku** [comment]
Hi @coreeey , pyroomacoustics uses a standard convolution (fftconvolve from scipy.signal) to convolve the source signal with the room impulse response.
Regardless of the function uses (scipy, numpy, speechbrain), the results should be the same up to some details.

The details that may change are:
1. The padding of the convolution. Scipy/numpy has mode same|valid|full depending on how you pad the input signal. The torch convolution used in speechbrain can certainly be padded to obtain the same result.
2. What speechbrain seems to do differently in `reverberate` is to align the time of arrival of the direct sound (the first/largest peak of the RIR) with time zero. This can be useful if you are creating data for single channel speech enhancement where you want to have time alignment between the noisy and clean signals. If you are doing multichannel processing, you should avoid to change the relative time origin of the RIR.

If you would like to use the RIR produced by pyroomacoustics with the reverb function from speechbrain (or any other) do as follows.
1. Call `room.compute_rir()`.
2. Access the rir between microphone `m` and source `s` in `room.rir[m][s]`.
3. Call `reverberate(my_signal, room.rir[m][s])`.

**Comment by coreeey** [comment]
Thank you, @fakufaku . When processing with multichannel systems, do the relative times of the RIRs reflect the time delays of different microphones?

**Comment by fakufaku** [comment]
Yes, the rir produced by the simulation have correct relative delays. They also have an extra 40 samples of global delay due to the fractional delay filters used to construct the impulse response. 

If you want to time align the rir without losing the relative delays, you should apply the same delay to all of them.
